packages/hierarchies-react/src/presentation-hierarchies-react/itwinui/TreeNodeRenderer.tsx

```diff
@@ -54,103 +62,136 @@ type TreeNodeRendererProps = Pick<ReturnType<typeof useTree>, "expandNode" | "is
-          {actions.slice(2, actions.length).map((action) => {
-            const info = action(node);
-            return (
-              <DropdownMenu.Item key={info.label} onClick={() => info.action()}>
-                {info.label}
-              </DropdownMenu.Item>
-            );
-          })}
-        </DropdownMenu.Content>
-      </DropdownMenu.Root>,
-    ];
-  };
+    const getActions = () => {
+      if (!actions || actions.length === 0) {
+        return undefined;
+      }
 
-  const selected = isNodeSelected(node.id);
-  return (
-    <Tree.Item
-      {...treeItemProps}
-      ref={nodeRef}
-      aria-level={node.level}
-      aria-posinset={node.posInLevel}
-      aria-setsize={node.levelSize}
-      label={getLabel ? getLabel(node) : node.label}
-      selected={selected}
-      expanded={node.isExpanded || node.children === true || node.children.length > 0 ? node.isExpanded : undefined}
-      onExpandedChange={(isExpanded) => {
-        expandNode(node.id, isExpanded);
-      }}
-      onClick={(event) => {
-        !treeItemProps["aria-disabled"] && onNodeClick?.(node, !selected, event);
-      }}
-      onKeyDown={(event) => {
-        // Ignore if it is called on the element inside, e.g. checkbox or expander
-        if (!treeItemProps["aria-disabled"] && event.target === nodeRef.current) {
-          onNodeKeyDown?.(node, !selected, event);
-        }
-      }}
-      icon={getIcon ? getIcon(node) : undefined}
-      actions={getActions()}
-    />
-  );
-}
+      if (actions.length < 4) {
+        return actions.map((action, index) => {
+          const actionInfo = action(node);
+          return <TreeActionButton key={index} {...actionInfo} />;
+        });
+      }
+
+      return [
+        <TreeActionButton key={0} {...actions[0](node)} />,
+        <TreeActionButton key={1} {...actions[1](node)} />,
+        <DropdownMenu.Root key={2}>
+          <DropdownMenu.Button render={<Tree.ItemAction icon={dropdownIcon} label="Tree actions dropdown" />} />
+          <DropdownMenu.Content>
+            {actions.slice(2, actions.length).map((action) => {
+              const info = action(node);
+              return (
+                <DropdownMenu.Item key={info.label} onClick={() => info.action()}>
+                  {info.label}
+                </DropdownMenu.Item>
+              );
+            })}
+          </DropdownMenu.Content>
+        </DropdownMenu.Root>,
+      ];
+    };
 
-function PlaceholderNode({
-  level,
-  ...props
-}: Omit<TreeNodeProps, "onExpanded" | "label" | "actions" | "aria-level" | "aria-posinset" | "aria-setsize" | "icon"> & { level: number }) {
+    const selected = isNodeSelected(node.id);
+    return (
+      <Tree.Item
+        {...treeItemProps}
+        ref={ref}
+        aria-level={node.level}
+        aria-posinset={node.posInLevel}
+        aria-setsize={node.levelSize}
+        label={getLabel ? getLabel(node) : node.label}
+        selected={selected}
```

Comment:
```suggestion
        selected={isNodeSelected(node.id)}
```

Reply:
Selected is used in multiple places i think it`s better to call it before hand.